### DISCUSSION TITLE
Expected behavior of definition conjunction inside default list?

### DISCUSSION THREAD

**Discussion by GUI** [cue-lang/cue#1648] [Q&A] [answered]
I understand that [this example](https://cuelang.org/play/?id=f1BevvPlezL#cue@export@json) should raise an error since `studio` is not part of the definition:

```cue
#Album: {
  artist: string
  title:  string
  year:   int
}

album: #Album & {
  artist: "Led Zeppelin"
  title:  "Led Zeppelin I"
  year:   1969
  studio: true
}
```

However, I'm trying to understand why [this example](https://cuelang.org/play/?id=uJirqury4jQ#cue@export@json) does not raise an error and instead just returns an empty list:

```cue
#Album: {
  artist: string
  title:  string
  year:   int
}

albums: [...#Album] | *[
  #Album & {
    artist: "Led Zeppelin"
    title:  "Led Zeppelin I"
    year:   1969
    studio: true
  }
]
```

[Without `studio: true`](https://cuelang.org/play/?id=Dz1mz5T9Rqf#cue@export@json), this default value seems to work as expected by providing a default album object inside of the list. But when the default value seems to be invalid, the result is empty, rather than an error.

I'm new to Cue, so if there's some other better way to achieve this, or or a different way to think about this, I'm open to suggestions. Or if this lack of error is expected, also curious about that. Basically, I'm trying to figure out how to have a list of structs that abide by a definition, while also providing a default value that abides by the same definition. Since these default values are static, I could obviously be more careful about ensuring they match the definition's schema, but I was hoping to be able to validate those default values in some way to make sure I don't make a mistake with those either. Thanks!

**Comment by NotTheEconomist** [comment, marked as answer]
You have a disjunction in your second example, so in English you're saying: "`albums` is a list of `#Album` structs _or_ this single-entry list I'm handing you as a default."

Turns out the single entry list can't unify and ends up as bottom, which means the disjunction falls back to the other definition which is just....an empty list of `#Album`s!

N.B. that if you replace your disjunction with a unify and drop the preference mark (the asterisk) then you get an error as you expect.

```
#Album: {
  artist: string
  title:  string
  year:   int
}

albums: [...#Album] & [
  {  // You don't even have to unify on #Album here since it's implied in the constraint
    artist: "Led Zeppelin"
    title:  "Led Zeppelin I"
    year:   1969
    studio: true  // This causes an error, failing to bottom
  }
]
```

**Reply by GUI** [reply]
@NotTheEconomist, thanks for the explanation! I'm afraid I may still not be fully grasping this, though. In your example, it does provide a validation error when that album is invalid, but the functionality is slightly different than what I was hoping to achieve. In your example, the `albums` value must be exactly that list of 1 album with no chance to provide a different value (eg, in another config file that's combined with this definition).

What I was trying to achieve was having that 1 album list only be the default value if no other list value was defined for this variable. So the default logic was important. And the default logic basically works as expected if the default value is valid, but I was just surprised that in the case the default value is invalid, no error is raised, so I was sort of wondering if this was a bug or intended behavior.

Put another way, this works as expected as long as `default_albums` is valid:

```cue
#Album: {
  artist: string
  title:  string
  year:   int
}

let default_albums = [
  #Album & {
    artist: "Led Zeppelin"
    title:  "Led Zeppelin I"
    year:   1969
  }
]
albums: [...#Album] | *default_albums
```

But when `default_albums` is made invalid, no error is raised, just an empty list is returned:

```cue
#Album: {
  artist: string
  title:  string
  year:   int
}

let default_albums = [
  #Album & {
    artist: "Led Zeppelin"
    title:  "Led Zeppelin I"
    year:   1969
    studio: true
  }
]
albums: [...#Album] | *default_albums
```

But this empty list behavior seems specific to the default operator preference operator, since without it, the same invalid definition of `default_albums` does raise an error (but in this case, `default_albums` is no longer really a default value, so this is basically what you proposed):

```cue
#Album: {
  artist: string
  title:  string
  year:   int
}

let default_albums = [
  #Album & {
    artist: "Led Zeppelin"
    title:  "Led Zeppelin I"
    year:   1969
    studio: true
  }
]
albums: default_albums
```

But perhaps I'm just basically restating what you already said, so sorry if that's the case. :) You say that since the default list can't be unified it falls back to an empty list, but I guess I'm just trying to understand if this is in fact expected behavior. Given how Cue validates and raises errors in most other contexts, I was just a little surprised that the default operator seems to change the behavior on these last two examples with one returning an en empty list and the other raising an error.

**Reply by eonpatapon** [reply]
You can make sure the default album list is valid by putting it in a regular field.

The way `cue` resolve disjuctions is by eliminating values that have an error. As long as there is one remaining valid value in a disjunction cue will not complain.

```cue
#Album: {
	artist: string
	title:  string
	year:   int
}

defaultAlbums: [...#Album]
defaultAlbums: [
	{
		artist: "Led Zeppelin"
		title:  "Led Zeppelin I"
		year:   1969
		// studio: true // uncomment to make cue raise an error
	},
]

albums: [...#Album] | *defaultAlbums
```

**Reply by GUI** [reply]
@eonpatapon, ah perfect, thank you for the suggestion! I think that's the behavior I was looking for! I did make one small tweak by changing `defaultAlbums` to a hidden field (`_defaultAlbums`), so that it isn't output in the final result, but then that seems to achieve what I was looking for. Thank you both!

```cue
#Album: {
	artist: string
	title:  string
	year:   int
}

_defaultAlbums: [...#Album]
_defaultAlbums: [
	{
		artist: "Led Zeppelin"
		title:  "Led Zeppelin I"
		year:   1969
		// studio: true // uncomment to make cue raise an error
	},
]

albums: [...#Album] | *_defaultAlbums
```

**Reply by verdverm** [reply]
Just passing by to say I ❤️  seeing the Led Zeppelin reference from Cuetorials
